hera_cal/nucal.py

```diff
@@ -1013,3 +1032,516 @@ def project_u_model_comps_on_spec_axis(u_model_comps, spectral_filters):
+        Array of real component of data with shape (Ntimes, Nbls)
+    data_i : np.ndarray
+        Array of imaginary component of data with shape (Ntimes, Nbls)
+    wgts : np.ndarray
+        Array of weights with shape (Ntimes, Nbls)
+    spectral_filters : np.ndarray
+        Array of spectral filters with shape (Nfreqs, Nfilters)
+    spatial_filters : List
+        List of spatial filters for each baseline in the group
+    idealized_blvecs : np.ndarray
+        Array of idealized baseline vectors with shape (Nbls, Ndims)
+
+    Returns:
+    -------
+    loss : float
+        Mean squared error between data and foreground model
+    """
+    # Compute foreground model from the model_parameters and DPSS filters
+    fg_model_r, fg_model_i = _foreground_model(model_parameters, spectral_filters, spatial_filters)
+
+    # Compute loss
+    return _mean_squared_error(model_parameters, data_r, data_i, wgts, fg_model_r, fg_model_i, idealized_blvecs)
+
+def _nucal_post_redcal(
+        data_r, data_i, wgts, model_parameters, optimizer, spectral_filters, spatial_filters, idealized_blvecs,
+        maxiter=100, convergence_criteria=1e-10, minor_cycle_maxiter=10
+    ):
+    """
+    Function to perform frequency redundant calibration using gradient descent. Calibrates the 
+    data by fitting a foreground model to the data using DPSS filters and estimating the redundant
+    calibration degeneracies that lead to the smoothest calibrated visibilities. Intended to only
+    be only after performing redundant calibration with redcal.RedundantCalibrator.
+
+    Parameters:
+    ----------
+    data_r : np.ndarray
+        Array of real component of data with shape (Ntimes, Nbls)
+    data_i : np.ndarray
+        Array of imaginary component of data with shape (Ntimes, Nbls)
+    wgts : np.ndarray
+        Array of weights with shape (Ntimes, Nbls)
+    model_parameters : dictionary
+        Parameters for fitting
+    optimizer : optax optimizer
+        Optimizer to use for gradient descent
+    spectral_filters : np.ndarray
+        Array of spectral filters with shape (Nfreqs, Nfilters)
+    spatial_filters : List
+        List of spatial filters for each baseline in the group
+    idealized_blvecs : np.ndarray
+        Array of idealized baseline vectors with shape (Nbls, Ndims)
+    maxiter : int, optional, default=100
+        Maximum number of iterations to perform
+    tol : float, optional, default=1e-10
+        Tolerance for stopping criterion. If the difference of the loss between two iterations is less than tol,
+        the optimization will stop.
+    minor_cycle_maxiter : int, optional, default=0
+        Maximum number of iterations of the minor cycle to perform after each major cycle. Minor cycles are performed by fixing the foreground
+        model and solving for the calibration parameters. When minor_cycle_maxiter is 0, no minor cycles are performed. If subsequent losses
+        are within tolerance, the minor cycle will stop.
```

Comment:
Clarify the relationship between maxiter and minor/major cycles

Reply:
Expanded on this